File: 17/17.cpp
#include "../leetcode.h"

class Solution {
public:
    // vector<string> res;memset
    char ch[5];
    char cd[5];
    int len;
    string dic[10] = {"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
    vector<string> res;
    vector<string> letterCombinations(string digits) {
        memset(ch,0,5*sizeof(char));
        len = digits.size();
        if (len == 0) return res;
        strcpy(cd, digits.c_str());
        // res = vector<string>(len*len*len*len);
        for (char c : dic[cd[0]-'0']){
            ch[0] = c;
            iter(0+1);
        }
        return res;
    };
    void iter(int ptr){
        if (ptr >= len) {
            res.push_back(string(ch));
            cout<<string(ch)<<endl;
            return;
        }
        for (char c : dic[cd[ptr]-'0']){
            ch[ptr] = c;
            iter(ptr+1);
        }
    }
};

int main(){
    Solution s = Solution();
    s.letterCombinations("5678");
    // s.letterCombinations("2");
    // s.letterCombinations("");
}

File: 17/17.java
import java.util.*;


class Solution {
    String [] dic = {"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
    char [] ch;
    String t;
    int len;
    List<String> res;
    public List<String> letterCombinations(String digits) {
        t = digits;
        len = t.length();
        res = new ArrayList();
        ch = new char[4];
        //  
        if (len == 0) return res;
        for (char c : dic[t.charAt(0)-'2'].toCharArray()){
            ch[0] = c;
            iter(1);
        }
        
        return res;
    };
    public void iter(int ptr){
        if (ptr >= len){
            String pre = new String(ch,0,ptr);
            res.add(pre);
            // System.out.println(pre);
            return;
        }
        for (char c : dic[t.charAt(ptr)-'2'].toCharArray()){
            ch[ptr] = c;
            iter(ptr+1);
        }
    };
}

class Main {

    public static void main(String[] args){  
        Solution s = new Solution();
        System.out.println(s.letterCombinations("23"));
        System.out.println(s.letterCombinations("5678"));
        System.out.println(s.letterCombinations(""));
        // String a = "asd";
        // a += 'b';
        // System.out.println(a);
        // char[] ch = {'g', 'o', 'o', 'd', 0, 'm', 'o', 'r', 'n', 'i', 'n', 'g'};
        // String str = new String(ch);
        // System.out.println(str);
    }
}
